pkg/registry/generic/rest/proxy.go

```diff
@@ -131,32 +138,42 @@ func (h *UpgradeAwareProxyHandler) tryUpgrade(w http.ResponseWriter, req *http.R
 		return false
 	}
 
-	backendConn, err := proxy.DialURL(h.Location, h.Transport)
+	var (
+		backendConn net.Conn
+		rawResponse []byte
+		err         error
+	)
+	if h.InterceptRedirects && utilconfig.DefaultFeatureGate.StreamingProxyRedirects() {
+		backendConn, rawResponse, err = h.connectBackendWithRedirects(req)
+	} else {
+		backendConn, err = h.connectBackend(req, h.Location)
+	}
 	if err != nil {
-		h.Responder.Error(err)
+		h.Responder.Error(fmt.Errorf("error dialing backend: %v", err))
 		return true
 	}
 	defer backendConn.Close()
 
-	requestHijackedConn, _, err := w.(http.Hijacker).Hijack()
-	if err != nil {
-		h.Responder.Error(err)
+	// Once the connection is hijacked, the ErrorResponder will no longer work, so
+	// hijacking should be the last step in the upgrade.
+	requestHijacker, ok := w.(http.Hijacker)
+	if !ok {
+		h.Responder.Error(fmt.Errorf("request connection cannot be hijacked: %T", w))
 		return true
 	}
-	defer requestHijackedConn.Close()
-
-	newReq, err := http.NewRequest(req.Method, h.Location.String(), req.Body)
+	requestHijackedConn, _, err := requestHijacker.Hijack()
 	if err != nil {
-		h.Responder.Error(err)
+		h.Responder.Error(fmt.Errorf("error hijacking request connection: %v", err))
 		return true
 	}
-	newReq.Header = req.Header
+	defer requestHijackedConn.Close()
 
-	if err = newReq.Write(backendConn); err != nil {
-		h.Responder.Error(err)
-		return true
+	// Forward raw response bytes back to client.
+	if _, err = requestHijackedConn.Write(rawResponse); err != nil {
+		glog.Errorf("Error proxying response from backend to client: %v", err)
```

Comment:
Done (What does this give us? Should I update the other error logging in the go routines below too?)